slkspec/core.py

```diff
@@ -135,23 +155,686 @@ def name(self) -> str:
+        # +----------------------------------------------------------
+        _start_recalls()
+        recall_timer = time.time()
+        # we do not generally remove files_recall_failed from to_be_retrieved because some files of failed recalls
+        # might have been recalled
+        while (
+            len(
+                [
+                    file_path
+                    for file_path in to_be_retrieved_files
+                    if file_path not in files_recall_failed.keys()
+                ]
+            )
+            > 0
+        ):
+            iterations += 1
+            retrieve_timer = time.time()
+            logger.info(
+                "retrieve/recall iteration %i; %i files to be retrieved. %i recall jobs running for %i files.",
+                iterations,
+                len(to_be_retrieved_files),
+                len(job_tape_mapping) + len(job_multi_tape_file_mapping),
+                sum(
+                    [len(tape_file_mapping[tape]) for tape in job_tape_mapping.values()]
+                )
+                + len(job_multi_tape_file_mapping),
+            )
+            for inp_file, out_dir in retrieve_files_corrected:
+                # skip files which do not need to be retrieved anymore
+                if inp_file not in to_be_retrieved_files:
+                    continue
+                Path(out_dir).mkdir(
+                    parents=True, exist_ok=True, mode=self.file_permissions
+                )
+                # check if recalls need to be started before retrieving
+                # check every 5 minutes whether additional recalls need to be started
+                if time.time() - recall_timer > 300:
+                    _start_recalls()
+                    recall_timer = time.time()
+                # check if file should be retrieved or not
+                output_dry_retrieve = pyslk.retrieve_improved(
+                    inp_file, out_dir, dry_run=True, preserve_path=False
+                )
+                # example output:
+                """
+                {'SKIPPED': {'SKIPPED_TARGET_EXISTS': ['/arch/bm0146/k204221/iow/INDEX.txt']},
+                    'FILES': {'/arch/bm0146/k204221/iow/INDEX.txt': '/home/k204221/tmp/INDEX.txt'}}
+
+                # dry run
+                {'ENVISAGED': {'ENVISAGED': ['/arch/bm0146/k204221/iow/INDEX.txt']},
+                    'FILES': {'/arch/bm0146/k204221/iow/INDEX.txt': '/home/k204221/tmp/abcdef2/INDEX.txt'}}
+
+                # after successful retrieval
+                {'ENVISAGED': {'ENVISAGED': []}, 'FILES': {'/arch/bm0146/k204221/iow/INDEX.txt':
+                    '/home/k204221/tmp/INDEX.txt'}, 'SUCCESS': {'SUCCESS': ['/arch/bm0146/k204221/iow/INDEX.txt']}}
+
+                {'FAILED': {'FAILED_NOT_CACHED': ['/arch/bm0146/k204221/iow/iow_data5_001.tar']},
+                    'FILES': {'/arch/bm0146/k204221/iow/iow_data5_001.tar': '/home/k204221/tmp/iow_data5_001.tar'}}
```

Comment:
This could go into the docstring of a new function?!

Reply:
I'll check.

Reply:
This docstring contains possible output of the function call four lines above. The following lines contain multiple if-clauses to react on the output. Therefore, I think it is reasonable to please this docstring here and not into the docstring of the function's header.